### PR TITLE
cpu: cleanup and optimize general structure

### PR DESCRIPTION
**Pull request by haukepetersen**
Currently, our general structure for CPUs and more importantly their include order/graph for their header files is inconsistent and not optimal. A simple PR like #6486 breaks, as the order and the structure of which includes are handled is not strictly hierarchical. To fix this, I propose the following:

- (1) get a little more hierarchical structure into our CPU folder by prefixing the folders like this:
```
$CPU  (e.g. `stm32f4`, `k64f`, ...)
   | base on
  v
$CPU_FAM (e.g. `fam_stm32`, `fam_kinetis`, ...)
  | based on
  v
$CPU_ARCH (e.g. `arch_cortexm`, `arch_msp430`, ...)
```
- (2) get the includes in order. Here I propose 2 'lines' of includes:
a: all 'atomic' (as in no other dependencies than into the C-lib)
```
cpu.h --> includes --> cpu_fam.h --> includes --> cpu_group.h
(in arch_xxx) --> (in fam_xxx) --> in (CPU)
```
where `cpu.h` is included by `external` modules. This 'line' of includes would always pull in all register definitions, all overriding of periph types etc.

b: all 'configuration' (CPU defines/types that have dependencies into atomic CPU types, but also into e.g. peripheral types)
```
cpu_conf.h --> includes --> cpu_conf_fam.h --> includes --> cpu_conf_group.h
(in arch_xxx) --> (in fam_xxx) --> in (CPU)
```

This would ensure, that the most specific values are always included first (and overriding more general values comes basically for free).

(1) is started in this PR (though not complete and untested. Needs more love...)

What do you think?

### CODE REVIEW

File: cpu/stm32f2/Makefile.include [hunk 1]
```diff
@@ -1,7 +1,7 @@
 export CPU_ARCH = cortex-m3
-export CPU_FAM  = stm32f2
+export CPU  = stm32f2
 
 USEMODULE += pm_layered
 
-include $(RIOTCPU)/stm32_common/Makefile.include
+include $(RIOTCPU)/fam_stm32/Makefile.includenclude
```
Comment (aabadie) [inline cpu/stm32f2/Makefile.include]: Makefile.include

**Comment by aabadie** [comment]
+1 in general with the idea you propose. One of the main improvement is the change from `CPU_FAM_<type>` to `CPU_<type>`. What I don't like is arch_ and fam_. Why not simply use a cpu/cortexm directory for common arm cortex m code and cpu/stm32, cpu/nrf5x, etc for common family code.

**Comment by kYc0o** [comment]
Can we maybe also put the hierarchy in the folder organisation? This is more or less what alexander proposes, since anyways all CPU's have a family and an arch, something like:

```
CPU ---> CPU_FAM ---> CPU_ARCH
```

So at the end you have e.g. :

```
cpu ---> stm32 ---> cortexm
```

Like this you don't need to prefix any names and the includes will also follow your proposed hierarchy.

Maybe the drawback is that we will have some duplicated folders (but not code) and will increase the path depth.

**Comment by miri64** [comment]
@haukepetersen With other restructurings of `cpu`: is this still valid and up to date?

**Comment by Teufelchen1** [comment]
@haukepetersen can this be closed? If not, speak up now - I am going to close this soonish™️

**Pull request closed without merging**